### PR TITLE
HAWQ-999. Treat hash table as random when file count is not in propor…

### PR DESCRIPTION
**Pull request by zhangh43**
…tion to bucket number of table.
By definition, file count of a hash table should be equal to or a multiple of the bucket number of the table. So if mismatch happens, we should not treat it as hash table in data locality algorithm.

### CODE REVIEW

**Comment by ztao1987** [comment]
+1

**Comment by paul-guo-** [comment]
+1

File: src/backend/cdb/cdbdatalocality.c [hunk 1]
```diff
@@ -3822,8 +3837,13 @@ run_allocation_algorithm(SplitAllocResult *result, List *virtual_segments, Query
...
 		/*allocate hash relation*/
 		if (isRelationHash) {
-			if (context->keep_hash && assignment_context.virtual_segment_num
-						== targetPolicy->bucketnum) {
+		  /*
+		   * if file count of the table is not equal to or multiple of
+		   * bucket number, we should process it as random table.
+		   */
+			if (context->keep_hash
+			    && assignment_context.virtual_segment_num== targetPolicy->bucketnum
```
Comment (hsyuan) [inline src/backend/cdb/cdbdatalocality.c]: space before ==

File: src/backend/cdb/cdbdatalocality.c [hunk 2]
```diff
@@ -3791,12 +3791,27 @@ run_allocation_algorithm(SplitAllocResult *result, List *virtual_segments, Query
 		targetPolicy = GpPolicyFetch(CurrentMemoryContext, myrelid);
 		bool isRelationHash = is_relation_hash(targetPolicy);
 
+		int fileCountInRelation = list_length(rel_data->files);
+		bool FileCountBucketNumMismatch = false;
```
Comment (hsyuan) [inline src/backend/cdb/cdbdatalocality.c]: why is this var capital leading? `FileCountBucketNumMismatch`

File: src/backend/cdb/cdbdatalocality.c [hunk 3]
```diff
@@ -3791,12 +3791,27 @@ run_allocation_algorithm(SplitAllocResult *result, List *virtual_segments, Query
...
+		if (targetPolicy->bucketnum > 0) {
+		  FileCountBucketNumMismatch = fileCountInRelation %
+		    targetPolicy->bucketnum == 0 ? false : true;
+		}
+		if (FileCountBucketNumMismatch && !allow_file_count_bucket_num_mismatch) {
+		  elog(ERROR, "file count %d in catalog is not in proportion to the bucket "
+		      "number %d of hash table with oid=%u, some data may be lost, if you "
```
Comment (hsyuan) [inline src/backend/cdb/cdbdatalocality.c]: number %d of distribution policy for table with oid = %u, some data may be missing or invalid.

File: src/backend/cdb/cdbdatalocality.c [hunk 4]
```diff
@@ -3791,12 +3791,27 @@ run_allocation_algorithm(SplitAllocResult *result, List *virtual_segments, Query
...
+		      "still want to continue the query by considering the table as random, set GUC "
```
Comment (hsyuan) [inline src/backend/cdb/cdbdatalocality.c]: You can force running the query by viewing the table as randomly distributed table. set GUC allow_file_count_bucket_num_mismatch on to enable this feature.

File: src/backend/utils/misc/guc.c [hunk 5]
```diff
@@ -450,6 +450,7 @@ char       *Debug_dtm_action_protocol_str;
 
 /* Enable check for compatibility of encoding and locale in createdb */
 bool		gp_encoding_check_locale_compatibility;
+bool  allow_file_count_bucket_num_mismatch;
```
Comment (hsyuan) [inline src/backend/utils/misc/guc.c]: indent

File: src/backend/utils/misc/guc.c [hunk 6]
```diff
@@ -3176,6 +3177,17 @@ static struct config_bool ConfigureNamesBool[] =
 	},
 
 	{
+	    {"allow_file_count_bucket_num_mismatch", PGC_POSTMASTER, CLIENT_CONN_LOCALE,
+	          gettext_noop("allow hash table to be treated as random when file count and"
```
Comment (hsyuan) [inline src/backend/utils/misc/guc.c]: allow file count and bucket number mismatch by viewing hash distributed table as randomly distributed table

File: src/backend/cdb/cdbdatalocality.c [hunk 7]
```diff
@@ -3791,12 +3791,27 @@ run_allocation_algorithm(SplitAllocResult *result, List *virtual_segments, Query
 		targetPolicy = GpPolicyFetch(CurrentMemoryContext, myrelid);
 		bool isRelationHash = is_relation_hash(targetPolicy);
 
+		int fileCountInRelation = list_length(rel_data->files);
+		bool FileCountBucketNumMismatch = false;
+		if (targetPolicy->bucketnum > 0) {
+		  FileCountBucketNumMismatch = fileCountInRelation %
```
Comment (hsyuan) [inline src/backend/cdb/cdbdatalocality.c]: why not `fileCountBucketNumMismatch = fileCountInRelation % targetPolicy->bucketnum != 0`

**Pull request closed without merging**